File: Project_and_Analysis_of_Algorithms/graph_search_applications/c++/bfs_distance.cpp
#include "graphs.h"

// bfsDistance returns the distance from the startNode to every other node in the given graph.
map<string, int> bfsDistance(adjacencyMap(string) graph, string startNode, visitedMap(string) visited){
    queue<string> q;
    q.push(startNode);
    visited[startNode] = true;

    map<string, int> distances; for (auto const&[node, _] : visited) distances[node] = -1;
    distances[startNode] = 0;

    while (!q.empty()) {
        string curNode = q.front(); q.pop();
        for (string neighbour : graph[curNode]) {
            if(!visited[neighbour]) {
                q.push(neighbour);
                visited[neighbour] = true;

                distances[neighbour] = distances[curNode] + 1;
            }
        }
    }

    return distances;
}

int main(void) {
    for(auto const&[node, distance] : bfsDistance(G1, "B", initVisited(G1))) cout << node << " : " << distance << "\n";
    return EXIT_SUCCESS;
}

File: Project_and_Analysis_of_Algorithms/graph_search_applications/python/bfs_distance.py
from collections import deque
from graphs import G1, init_visited

def bfs_distance(graph: dict[str, list[str]], start_node: str, visited: dict[str, bool]) -> dict[str, int]:
    """Returns the distances from start_node to every other node in the graph using BFS."""
    queue = deque([start_node])
    visited[start_node] = True
    distances: dict[str, int] = visited.copy()
    distances[start_node] = 0

    while len(queue) > 0:
        cur_node: str = queue.popleft()
        for neighbour in graph[cur_node]:
            if not visited[neighbour]:
                queue.append(neighbour)
                visited[neighbour] = True
                distances[neighbour] = distances[cur_node] + 1

    return distances


def main():
    print(bfs_distance(graph=G1, start_node="B", visited=init_visited(graph=G1)))
    return


if __name__ == "__main__":
    main()
